### PR TITLE
[Merged by Bors] - Reuse texture when resolving multiple passes

### PR DESCRIPTION
**Pull request by HackerFoo**
# Objective

Fixes https://github.com/bevyengine/bevy/issues/3499

## Solution

Uses a `HashMap` from `RenderTarget` to sampled textures when preparing `ViewTarget`s to ensure that two passes with the same render target get sampled to the same texture.

This builds on and depends on https://github.com/bevyengine/bevy/pull/3412, so this will be a draft PR until #3412 is merged. All changes for this PR are in the last commit.

### CODE REVIEW

**Comment by HackerFoo** [comment]
It looks like the depth buffer is also not being shared by passes.

**Comment by superdump** [comment]
As #3412 was merged, could you update this on top of main so we can get it reviewed and merged?

File: crates/bevy_core_pipeline/src/main_pass_3d.rs [hunk 1]
```diff
@@ -139,7 +139,7 @@ impl Node for MainPass3dNode {
                     // transparent ones.
                     depth_ops: Some(Operations {
                         load: LoadOp::Load,
-                        store: false,
+                        store: true,
```
Comment (superdump) [inline crates/bevy_core_pipeline/src/main_pass_3d.rs:142]: Please update the preceding comment and explain why the depth buffer is being written to as it isn't obvious from this code alone, and Xcode will likely warn about it as wasted bandwidth usage so people may come here and want to change it. If there is a comment explaining why it is being written, then we can hopefully avoid some ping-ponging. :)

**Review by superdump — CHANGES_REQUESTED** [review]
Just need to update a comment and then this looks good to me.

File: crates/bevy_core_pipeline/src/main_pass_3d.rs (hunk 1 above)
Comment (HackerFoo) [reply crates/bevy_core_pipeline/src/main_pass_3d.rs:144]: Done. I also filed #4174 as a reminder to remove the workaround after the issue is resolved.

File: crates/bevy_core_pipeline/src/main_pass_3d.rs [hunk 2]
```diff
@@ -134,12 +134,14 @@ impl Node for MainPass3dNode {
                 })],
                 depth_stencil_attachment: Some(RenderPassDepthStencilAttachment {
                     view: &depth.view,
-                    // NOTE: For the transparent pass we load the depth buffer but do not write to it.
+                    // NOTE: For the transparent pass we load the depth buffer. There should be no
+                    // need to write to it, but store is set to `true` as a workaround for issue #3776,
```
Comment (alice-i-cecile) [inline crates/bevy_core_pipeline/src/main_pass_3d.rs:138]: Please include a URL link here: it's much more convenient.

File: examples/3d/two_passes.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+/// Rotates the outer cube (main pass)
+fn cube_rotator_system(time: Res<Time>, mut query: Query<&mut Transform, With<MainPassCube>>) {
+    for mut transform in query.iter_mut() {
+        transform.rotation *= Quat::from_rotation_x(1.0 * time.delta_seconds());
+        transform.rotation *= Quat::from_rotation_y(0.7 * time.delta_seconds());
+    }
+}
+
+fn cycle_msaa(input: Res<Input<KeyCode>>, mut msaa: ResMut<Msaa>) {
```
Comment (alice-i-cecile) [inline examples/3d/two_passes.rs:207]:
```suggestion
fn toggle_msaa(input: Res<Input<KeyCode>>, mut msaa: ResMut<Msaa>) {
```

File: examples/3d/two_passes.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+fn main() {
+    let mut app = App::new();
+    app.insert_resource(Msaa { samples: 4 })
+        .add_plugins(DefaultPlugins)
+        .add_plugin(CameraTypePlugin::<FirstPassCamera>::default())
+        .add_startup_system(setup)
+        .add_system(cube_rotator_system)
+        .add_system(rotator_system)
+        .add_system(cycle_msaa);
```
Comment (alice-i-cecile) [inline examples/3d/two_passes.rs:30]:
```suggestion
        .add_system(toggle_msaa);
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
A couple of nits, but the example and general code quality look good to me. Cannot comment in depth on the rendering strategy.

File: crates/bevy_core_pipeline/src/main_pass_3d.rs (hunk 2 above)
Comment (HackerFoo) [reply crates/bevy_core_pipeline/src/main_pass_3d.rs:146]: Done.

File: examples/3d/two_passes.rs (hunk 4 above)
Comment (HackerFoo) [reply examples/3d/two_passes.rs]: Done.

File: examples/3d/two_passes.rs (hunk 3 above)
Comment (HackerFoo) [reply examples/3d/two_passes.rs]: Done.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by cart** [comment]
Sorry for the delay on this. This seems relatively uncontroversial / worth including in 0.7. I'll try to give it a proper review before release.

**Pull request closed without merging**